### PR TITLE
Template MemoryPool::get_usage_statistics and print_state

### LINKED ISSUE #7555: Nightly build failures, gcc/{11.3.0, 12.2.0, 13.2.0}, OpenMP+Serial backend: Kokkos_MemoryPool.hpp:169:45: error: invalid use of incomplete type 'using DefaultHostExecutionSpace = class Kokkos::OpenMP'
**Describe the bug**

A nightly Trilinos job with gcc/11.3.0 and the OpenMP backend enabled is failing to compile with output:
```
02:43:36 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_TaskQueue.hpp:33,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial_Task.hpp:27,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial.hpp:287,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_Default_Graph_Impl.hpp:26,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Graph.hpp:219,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp: In member function 'void Kokkos::MemoryPool<Device>::get_usage_statistics(Kokkos::MemoryPool<Device>::usage_statistics&) const':
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp:169:45: error: invalid use of incomplete type 'using DefaultHostExecutionSpace = class Kokkos::OpenMP' {aka 'class Kokkos::OpenMP'}
02:43:36   169 |           Kokkos::DefaultHostExecutionSpace{}, sb_state_array, m_sb_state_array,
02:43:36       |                                             ^
02:43:36 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Build/kokkos/KokkosCore_Config_FwdBackend.hpp:22,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Core_fwd.hpp:79,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_GraphImpl.hpp:22,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Graph.hpp:31,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/fwd/Kokkos_Fwd_OPENMP.hpp:22:7: note: forward declaration of 'using DefaultHostExecutionSpace = class Kokkos::OpenMP' {aka 'class Kokkos::OpenMP'}
02:43:36    22 | class OpenMP;  ///< OpenMP execution space.
02:43:36       |       ^~~~~~
02:43:36 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_TaskQueue.hpp:33,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial_Task.hpp:27,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial.hpp:287,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_Default_Graph_Impl.hpp:26,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Graph.hpp:219,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp: In member function 'void Kokkos::MemoryPool<Device>::print_state(std::ostream&) const':
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp:222:45: error: invalid use of incomplete type 'using DefaultHostExecutionSpace = class Kokkos::OpenMP' {aka 'class Kokkos::OpenMP'}
...
```

This follows after merge of these changes:

**Changes**

    Print value of all relevant SYCL_EXT_ONEAPI* macros ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/440/changes#detail))
    Always pass execution space to `Impl::DeepCopy` (#7519) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/440/changes#detail))

Presumably correlated with #7519

I'll test a similar standalone Kokkos build to see about reproducing outside of Trilinos

**Please include the following for a minimal reproducer**

**Reproducer (with Trilinos)**: (blake all queue)
```bash
# Repo prep
git clone -b develop https://github.com/trilinos/Trilinos.git
TRILINOS_DIR=$PWD/Trilinos
git clone -b develop https://github.com/kokkos/kokkos.git
KOKKOS_DIR=$PWD/kokkos
git clone -b develop https://github.com/kokkos/kokkos-kernels.git
KOKKOSKERNELS_DIR=$PWD/kokkos-kernels

cd $TRILINOS_DIR
ln -s ${KOKKOS_DIR}/kokkos kokkos
ln -s ${KOKKOSKERNELS_DIR}/kokkos-kernels kokkos-kernels
cd -

mkdir -p Build
cd Build


salloc -N 1 -p all

# Environment
module purge
module load cmake gcc/11.3.0 openmpi/4.1.5 openblas/0.3.23

# Configure Trilinos - likely only Kokkos needs to be enabled
cmake \
  -D CMAKE_CXX_COMPILER="`which mpicxx`" \
  -D CMAKE_C_COMPILER="`which mpicc`" \
  -D CMAKE_Fortran_COMPILER="`which mpifort`" \
  -D CMAKE_CXX_FLAGS="-g" \
  -D CMAKE_C_FLAGS="-g" \
  -DTPL_ENABLE_MPI=ON \
  -DTPL_ENABLE_BLAS:BOOL=ON \
   -DBLAS_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DBLAS_LIBRARY_NAMES:STRING="openblas" \
  -DTPL_ENABLE_LAPACK:BOOL=ON \
   -DLAPACK_INCLUDE_DIRS:FILEPATH=${OPENBLAS_ROOT}/include \
   -DLAPACK_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DLAPACK_LIBRARY_NAMES:STRING="openblas" \
  -DTrilinos_ENABLE_ALL_PACKAGES=OFF \
  -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
  -DTrilinos_ENABLE_TESTS=ON \
  -DTrilinos_MUST_FIND_ALL_TPL_LIBS=TRUE \
  -DTrilinos_ENABLE_COMPLEX=ON \
  -DTrilinos_ENABLE_OpenMP=ON \
  -DTrilinos_ENABLE_Kokkos=ON \
  -D Kokkos_ENABLE_OPENMP=ON \
   -D Kokkos_ENABLE_TESTS=ON \
  -D Kokkos_ARCH_SKX=ON \
   -D Kokkos_ENABLE_IMPL_MDSPAN=ON \
  -DTrilinos_ENABLE_KokkosKernels=ON \
   -D KokkosKernels_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Tpetra=ON \
   -D Tpetra_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Sacado=ON \
   -D Sacado_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Stokhos=ON \
   -D Stokhos_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Intrepid2=ON \
   -D Intrepid2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Amesos2=ON \
   -D Amesos2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Ifpack2=ON \
   -D Ifpack2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Zoltan2=ON \
   -D Zoltan2_ENABLE_TESTS=ON \
\
  -DTPL_ENABLE_Matio=OFF \
\
-DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
-DKokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
$TRILINOS_DIR
```

Reproducer (kokkos only):
```bash
module load cmake gcc/11.3.0 openmpi/4.1.5 openblas/0.3.23

cmake \
  -D CMAKE_CXX_COMPILER="g++" \
  -D CMAKE_CXX_FLAGS="-g" \
  -D CMAKE_BUILD_TYPE="RELEASE" \
  -D Kokkos_ENABLE_SERIAL=ON \
  -D Kokkos_ENABLE_OPENMP=ON \
  -D Kokkos_ENABLE_TESTS=ON \
  -D Kokkos_ARCH_SKX=ON \
  -D Kokkos_ENABLE_IMPL_MDSPAN=ON \
$KOKKOS_DIR
```

* The explicit enable of both the OPENMP and SERIAL backends was needed to trigger the error; only enabling the OPENMP backend allowed compilation to succeed

Edit: added standalone kokkos reproducer infor

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/7555.

### CODE REVIEW

**Comment by ndellingwood** [comment]
@masterleinad should an update be made e.g. to the GCC-8.4.0 job to add coverage for OpenMP + Serial?
```diff
diff --git a/.jenkins b/.jenkins
index 5cccc2202..42fe14d7e 100644
--- a/.jenkins
+++ b/.jenkins
@@ -58,6 +58,7 @@ pipeline {
                                 -DKokkos_ENABLE_TESTS=ON \
                                 -DKokkos_ENABLE_BENCHMARKS=ON \
                                 -DKokkos_ENABLE_OPENMP=ON \
+                                -DKokkos_ENABLE_SERIAL=ON \
                                 -DKokkos_ENABLE_LIBDL=OFF \
                                 -DKokkos_ENABLE_LIBQUADMATH=ON \
                                 -DKokkos_ENABLE_SERIAL=ON \
```

**Review by ndellingwood — APPROVED** [review]

**Comment by masterleinad** [comment]
> @masterleinad should an update be made e.g. to the GCC-8.4.0 job to add coverage for OpenMP + Serial?

That configuration is already enabling the `Serial` backend explicitly just a couple of lines lower.

**Comment by ndellingwood** [comment]
@masterleinad oop, I overlooked that. I wonder why the issue did not show up with that job?

File: core/src/Kokkos_MemoryPool.hpp [hunk 1]
```diff
@@ -156,8 +156,13 @@ class MemoryPool {
     size_t reserved_bytes;   ///<  Unallocated bytes in assigned superblocks
   };
 
+  // This function is templated to avoid needing a full definition of
+  // DefaultHostExecutionSpace at class instantiation
+  template <typename ExecutionSpace = Kokkos::DefaultHostExecutionSpace>
```
Comment (nmm0) [inline core/src/Kokkos_MemoryPool.hpp:161]: Can't we move the definition outside of the class and place it after DefaultHostExecutionSpace is a complete type?

Comment (masterleinad) [reply core/src/Kokkos_MemoryPool.hpp:161]: The class is templated so we can't put it in a source file without explicit template instantiations. I'm not sure that creating a separate header file would be preferrable over the solution here.

**Comment by dalg24** [comment]
> > @masterleinad should an update be made e.g. to the GCC-8.4.0 job to add coverage for OpenMP + Serial?
> 
> That configuration is already enabling the `Serial` backend explicitly just a couple of lines lower.

How do we explain that the CI did not catch the issue before?

**Comment by masterleinad** [comment]
> How do we explain that the CI did not catch the issue before?

Not sure yet. The `HPX` build also explicitly enables `Serial`.

File: core/src/Kokkos_MemoryPool.hpp (hunk 1 above)
Comment (nmm0) [reply core/src/Kokkos_MemoryPool.hpp:161]: I just don't like templating the function and then static_asserting on the type. I don't see any problem with having an impl header

**Comment by pzehner** [comment]
I get the original error on `develop` for the target `Kokkos_HeaderSelfContained_Kokkos_Graph_hpp` when both `Kokkos_ENABLE_OPENMP`, `Kokkos_ENABLE_TESTS` and `Kokkos_ENABLE_HEADER_SELF_CONTAINMENT_TESTS` are enabled. The latest option is automatically enabled when tests are enabled, so I don't understand why the CI didn't catch this.

File: core/src/Kokkos_MemoryPool.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/Kokkos_MemoryPool.hpp:161]: Note that we are already doing this in https://github.com/kokkos/kokkos/pull/7519 (which caused the issue) in other places. I just don't want to figure out the appropriate include order to make all of this work. Feel free to give it a shot if you have a better solution!

**Comment by pzehner** [comment]
I dug a bit more and found that the error appears starting with GCC 9.4.0.

**Comment by pzehner** [comment]
It would be nice to test this somehow.

**Comment by masterleinad** [comment]
> It's hard to test every single configuration we have seen failing. We could try enabling `Serial` for all(?) GitHub workflow builds if that's not too expensive.

See https://github.com/kokkos/kokkos/actions/runs/12238747792/job/34141364479?pr=7587 for tests that would have caught the problem.

**Comment by pzehner** [comment]
While I recognize it's not possible to test each and every combination of compiler, version, and options, I think this case should be covered, because it's a practical case. I'm wondering if testing with both `Kokkos_ENABLE_SERIAL` and `Kokkos_ENABLE_OPENMP` instead of with just `Kokkos_ENABLE_OPENMP` won't have other effects.

**Comment by masterleinad** [comment]
@pzehner Feel free to propose a different solution if you don't agree with either not adding builds or adding `Serial` to the GitHub workflow builds.

**Comment by JBludau** [comment]
I kind of feel like we should absolutely ensure that `Serial` is always working. Nevertheless, I am not sure if adding it is to much additional strain on the CI. 

> I'm wondering if testing with both `Kokkos_ENABLE_SERIAL` and `Kokkos_ENABLE_OPENMP` instead of with just `Kokkos_ENABLE_OPENMP` won't have other effects.

Can we try this and see if we get swamped by issues?

**Comment by masterleinad** [comment]
> I kind of feel like we should absolutely ensure that `Serial` is always working. Nevertheless, I am not sure if adding it is to much additional strain on the CI.

The last commit makes all the builds like 10 min slower (50min -> 60min or so).

> Can we try this in a separate setting and see if we get swamped by issues?

Can you elaborate? In which cases, would you expect more issues?

> But maybe we should track this in an issue and keep this PR about the fix

I'd agree with that.

File: .github/workflows/continuous-integration-linux.yml [hunk 2]
```diff
@@ -118,6 +118,7 @@ jobs:
             -DKokkos_ENABLE_DESUL_ATOMICS_EXTERNAL=ON \
             -DKokkos_ENABLE_HWLOC=ON \
             -DKokkos_ENABLE_${{ matrix.backend }}=ON \
+            -DKokkos_ENABLE_SERIAL=ON \
```
Comment (dalg24) [inline .github/workflows/continuous-integration-linux.yml:121]: Please drop that change
I'd rather have no coverage in CI that enabling the Serial backend in all these builds

Comment (masterleinad) [reply .github/workflows/continuous-integration-linux.yml]: Dropped. That was my preference as well.

**Review by dalg24 — APPROVED** [review]
Please make sure we keep track of testing elsewhere

**Pull request merged**